### PR TITLE
Allow prefix to be specified per source (#2650)

### PR DESCRIPTION
**Pull request by amarjandu**
Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented on demo expectations                        <sub>or labelled PR as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that Demo expectations are clear              <sub>or PR is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and build passes in `sandbox` <sub>or added `no sandbox` label</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Purchased BQ slot committment in `dev`                <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Deleted BQ slot committment in `dev`                  <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Purchased BQ slot committment in `prod`               <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Deleted BQ slot committment in `prod`                 <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

**Comment by amarjandu** [comment]
Testing will fail;  env vars need to be updated if we decide to go with this change.
Waiting on https://github.com/DataBiosphere/azul/issues/2650#issuecomment-784402746

**Comment by amarjandu** [comment]
Need to update the `scripts/reindex.py` and inspect its usage in pipelines.

File: UPGRADING.rst [hunk 1]
```diff
@@ -11,6 +11,14 @@ reverted. This is all fairly informal and loosely defined. Hopefully we won't
 have too many entries in this file.
 
 
+#2650 Add prefix to catalogs
+============================
+
+The syntax of the value of the AZUL_CATALOGS environment variable was modified
+to include a prefix value. To upgrade a deployment, append every catalog entry
+in that variable with a prefix ``:foo``.
```
Comment (hannes-ucsc) [inline UPGRADING.rst:19]: Prefixes have to be hexadecimal so this isn't a representative example.

File: src/azul/__init__.py [hunk 2]
```diff
@@ -526,13 +527,17 @@ def catalogs(self) -> Mapping[CatalogName, Catalog]:
...
-            for atlas, catalog, *plugins in map(splitter(':'), catalogs)
+            for atlas, catalog, *plugins, prefix in map(splitter(':'), catalogs)
         }
         require(bool(catalogs), 'No catalogs configured')
-        reject(any('/' in catalog_name for catalog_name in catalogs),
-               'It appears AZUL_CATALOGS was not upgraded to include atlas names.')
+        for catalog_name, catalog in catalogs.items():
+            reject('/' in catalog_name,
+                   'It appears AZUL_CATALOGS was not upgraded to include atlas names.')
+            reject(catalog.prefix == '')
```
Comment (hannes-ucsc) [inline src/azul/__init__.py:539]: Empty prefixes should be accepted. In fact, they are the norm.

**Review by hannes-ucsc — COMMENTED** [review]
Unfortunately, the spec in the ticket has been obsoleted by recent changes and I failed to keep it up to date. The prefix should be specific to the source, not the catalog. I'll update the ticket.

**Comment by hannes-ucsc** [comment]
@amarjandu note my edit to the description of the underlying ticket.

File: src/azul/azulclient.py [hunk 3]
```diff
@@ -262,8 +259,8 @@ def remote_reindex_partition(self, message: JSON) -> None:
         bundle_fqids = self.list_bundles(catalog, source, prefix)
         bundle_fqids = self.filter_obsolete_bundle_versions(bundle_fqids)
         logger.info('After filtering obsolete versions, '
-                    '%i bundles remain in prefix %r of catalog %r',
-                    len(bundle_fqids), prefix, catalog)
+                    '%i bundles remain in prefix %r for the source %r in catalog %r',
```
Comment (hannes-ucsc) [inline src/azul/azulclient.py:262]:
```suggestion
                    '%i bundles remain in prefix %r of source %r in catalog %r',
```

File: src/azul/azulclient.py [hunk 4]
```diff
@@ -282,7 +279,8 @@ def remote_reindex_partition(self, message: JSON) -> None:
             ]
             self.notifications_queue.send_messages(Entries=entries)
             num_messages += len(batch)
-        logger.info('Successfully queued %i notification(s) for prefix %s', num_messages, prefix)
+        logger.info('Successfully queued %i notification(s) for prefix %s for '
+                    'the source %r', num_messages, prefix, source)
```
Comment (hannes-ucsc) [inline src/azul/azulclient.py:283]:
```suggestion
        logger.info('Successfully queued %i notification(s) for prefix %s of '
                    'source %r', num_messages, prefix, source)
```

File: src/azul/terra.py [hunk 5]
```diff
@@ -75,6 +78,7 @@ class TDRSourceName(SourceName):
     project: str
     name: str
     is_snapshot: bool
+    prefix: str
```
Comment (hannes-ucsc) [inline src/azul/terra.py:81]:
```suggestion
    prefix: str = ''
```

File: test/indexer/test_tdr.py [hunk 6]
```diff
@@ -62,7 +62,8 @@ class TestTDRPlugin(CannedBundleTestCase):
     source = TDRSourceRef(id='test_id',
                           name=TDRSourceName(project='test_project',
                                              name='snapshot',
-                                             is_snapshot=True))
+                                             is_snapshot=True,
+                                             prefix=''))
```
Comment (hannes-ucsc) [inline test/indexer/test_tdr.py:66]:
```suggestion
                                             is_snapshot=True))
```

File: scripts/reindex.py [hunk 7]
```diff
@@ -107,9 +101,9 @@ def main(argv: List[str]):
         logger.info('Queuing notifications for reindexing ...')
         for catalog in args.catalogs:
             if args.partition_prefix_length:
-                azul.remote_reindex(catalog, args.prefix, args.partition_prefix_length)
+                azul.remote_reindex(catalog, args.partition_prefix_length)
             else:
-                azul.reindex(catalog, args.prefix)
+                azul.reindex(catalog, '')
```
Comment (hannes-ucsc) [inline scripts/reindex.py:106]: This is the only call site so it doesn't make sense to fix this here but retain the argument. Why did you remove the args.reindex option?

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: scripts/reindex.py (hunk 7 above)
Comment (amarjandu) [reply scripts/reindex.py:106]: I think you are still reviewing this, but did you mean `args.prefix`?

Comment (hannes-ucsc) [reply scripts/reindex.py:106]: Uhm, yes. Good catch.

File: deployments/sandbox/environment.py [hunk 8]
```diff
@@ -94,7 +96,7 @@ def env() -> Mapping[str, Optional[str]]:
             }
         ),
 
-        'azul_dss_query_prefix': '42',
+        'AZUL_DSS_QUERY_PREFIX': '42',
```
[the remark this answers is not in the document]
Comment (amarjandu) [reply deployments/sandbox/environment.py:99]: I did not think it was necessary to mention this change because it would only be used if someone was interfacing with the DSS.
Azul.config also provides default value `''` if it does not locate the env. var. https://github.com/DataBiosphere/azul/blob/8f15f2020f413379b80a57793651cc54ac7e3d7f/src/azul/__init__.py#L202


If you feel that we should include a section about this change I can totally do so, my goal keep this focused on the steps that we actually need.

[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply deployments/sandbox/environment.py:99]: > because it would only be used if someone was interfacing with the DSS

Not quite. On `develop` the variable is used for TDR as well, despite the name. People may have this set in their personal deployments. If they do, they should remove the variable from the environment, *unless* they plan to interface with a DSS instance in the future, which is unlikely.

File: scripts/reindex.py (hunk 7 above)
Comment (amarjandu) [reply scripts/reindex.py:106]: I removed the `args.prefix`option because it seemed like it would not be needed if we hardcode the prefix to the source. By allowing a prefix to be provided to `azulclient.reindex` the reindexed bundles would be those under `{source.name.prefix} + {args.prefix}`. At the time I was thinking that if we are required to redeploy to update the the source prefix, why not just make this the only way to specify the prefix.

Maybe it's better to retain the args.prefix, make the default value of it `''`, and update the description to mention that its used to narrow down the source prefix, rather than to reindex an entirely different prefix.

> This is the only call site so it doesn't make sense to fix this here but retain the argument.

I would think that we want the azulclient to be robust enough to handle a prefix during reindex, so I opted to retain the prefix argument in that method vs removing it, even though its only used in this one call site.

I think there is a problem that we overload the term `prefix` in the repository plugins to mean `source.name.prefix` or `partition prefix`. By having distinct optional arguments for prefix and partition prefix within the methods of the repository plugin we can make make the client more flexible, and is probably something worth looking into in the long run.

**Comment by amarjandu** [comment]
pulling from review to rebase

**Comment by amarjandu** [comment]
unassigning to address conflicts to reindexing script.

File: UPGRADING.rst [hunk 9]
```diff
@@ -11,6 +11,20 @@ reverted. This is all fairly informal and loosely defined. Hopefully we won't
...
+===========================
+
+For non-main deployments unset the ``azul_dss_query_prefix`` variable.
+
+The syntax of ``AZUL_TDR_SOURCES`` and ``AZUL_TDR_…_SOURCES`` environment
+variables was modified to include a UUID prefix. To upgrade a
+deployment, append every source entry in the deployment's `environment.py`
+with a colon delimiter ``:`` followed by a valid hexadecimal prefix e.g.,
+``:42``. When selecting a prefix for an IT snapshot source on a non-main
+deployment, use an empty string, failure to do so may cause IT errors.
```
Comment (hannes-ucsc) [inline UPGRADING.rst:24]:
```suggestion
deployment, use an empty string. Failure to do so may cause IT errors.
```

File: scripts/reindex.py [hunk 10]
```diff
@@ -45,10 +48,10 @@
...
                     help='The length of the bundle UUID prefix by which to partition the set of bundles matching the '
                          'query. Each query partition is processed independently and remotely by the indexer lambda. '
-                         'The lambda queries the DSS and queues a notification for each matching bundle. If 0 (the '
-                         'default) no partitioning occurs, the DSS is queried locally and the indexer notification '
-                         'endpoint is invoked for each bundle individually and concurrently using worker threads. '
-                         'This is magnitudes slower that partitioned indexing.')
+                         'The lambda queries the repository and queues a notification for each matching bundle. If 0 '
+                         '(the default) no partitioning occurs, the repository is queried locally and the indexer '
+                         'notification endpoint is invoked for each bundle individually and concurrently using worker'
+                         'threads. This is magnitudes slower that partitioned indexing.')
```
Comment (hannes-ucsc) [inline scripts/reindex.py:54]:
```suggestion
                         'threads. This is magnitudes slower than remote i.e., partitioned indexing.')
```

File: src/azul/plugins/repository/dss/__init__.py [hunk 11]
```diff
@@ -93,11 +96,17 @@ def lookup_source_id(self, name: SimpleSourceName) -> str:
     def dss_client(self):
         return client(dss_endpoint=config.dss_endpoint)
 
+    @cached_property
```
Comment (hannes-ucsc) [inline src/azul/plugins/repository/dss/__init__.py:99]: `config.dss_query_prefix` is cheap so there is no point in caching it. OTOH, doing so only makes mocking harder.

Last but not least, there is only one reference to this so even a normal property seems like overkill.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Comment by hannes-ucsc** [comment]
And note spelling errors in commit titles, @amarjandu.

File: UPGRADING.rst [hunk 12]
```diff
@@ -11,6 +11,20 @@ reverted. This is all fairly informal and loosely defined. Hopefully we won't
...
+#2650 Add prefix to sources
+===========================
+
+For non-main deployments unset the ``azul_dss_query_prefix`` variable.
+
+The syntax of ``AZUL_TDR_SOURCES`` and ``AZUL_TDR_…_SOURCES`` environment
+variables was modified to include a UUID prefix. To upgrade a
+deployment, append every source entry in the deployment's `environment.py`
+with a colon delimiter ``:`` followed by a valid hexadecimal prefix e.g.,
+``:42``. When selecting a prefix for an IT snapshot source on a non-main
```
[the remark this answers is not in the document]
Comment (amarjandu) [reply UPGRADING.rst]: I used `non-main` here because the IT catalogs within the `sandbox` also need to have a empty string prefix to allow the IT to run correctly.

[the remark this answers is not in the document]
Comment (amarjandu) [reply UPGRADING.rst]: Perhaps we can consider the sandbox to be a shared personal deployment.

[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply UPGRADING.rst]: You're upgrading the sandbox in this PR, correct? If so, there is nothing else to do for the sandbox.

**Comment by amarjandu** [comment]
Removing from review to resolve conflicts.

File: deployments/sandbox/environment.py [hunk 13]
```diff
@@ -53,19 +53,21 @@ def env() -> Mapping[str, Optional[str]]:
...
+                'tdr:broad-jade-dev-data:snapshot/hca_dev_20201023_ebiv4___20210302:42'
+            ]),
+            f'AZUL_TDR_{"it2ebi".upper()}_SOURCES': ','.join([
+                'tdr:broad-jade-dev-data:snapshot/hca_dev_20201023_ebiv4___20210302:'
+            ]),
+            f'AZUL_TDR_{"lungmap".upper()}_SOURCES': ','.join([
+                'tdr:broad-jade-dev-data:snapshot/lungmap_dev_20210203___20210204:42',
+            ]),
+            f'AZUL_TDR_{"it3lungmap".upper()}_SOURCES': ','.join([
+                'tdr:broad-jade-dev-data:snapshot/lungmap_dev_20210203___20210204:',
```
Comment (hannes-ucsc) [inline deployments/sandbox/environment.py:69]: There is no point in using ** on a literal dict, the `"dcp2ebi".upper()` is convoluted and each snapshot is mentioned twice.

File: scripts/recan_bundle_tdr.py [hunk 14]
```diff
@@ -387,7 +387,7 @@ def main(argv):
         metadata = json.load(f)
 
     dss_source = DSSSourceRef(id='',
-                              name=SimpleSourceName(''))
+                              name=DSSSourceName(''))
```
Comment (hannes-ucsc) [inline scripts/recan_bundle_tdr.py:390]: This can't work because `DSSSourceName` uses `kwonly=True`. PyCharm warns me about this. Please make sure you get a warning too and reply with a screenshot of that warning before addressing this. I want to make sure you use the right PyCharm inspection configuration.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Conflicts. Last time there were conflicts, it appears that the fixups were squashed out. Ideally, that shouldn't be done.

I don't think the commit "Add prefix to SourceName" should be separate. It isn't self-contained, likely to be reverted by itself or makes reviewing the PR easier. It's title isn't accurate either: the commit does not add a prefix to `SourceName`, it doesn't touch `SourceName` at all.

However, since all sub-classes of `SourceName` have `prefix`, we *should* actually pull it up into the base class. Since this is a fairly large change I don't think I need to see a fixup in the next round.

File: scripts/recan_bundle_tdr.py (hunk 14 above)
Comment (amarjandu) [reply scripts/recan_bundle_tdr.py:390]: Screen Shot 2021-04-14 at 2 00 47 PM

File: src/azul/plugins/repository/dss/__init__.py [hunk 15]
```diff
@@ -57,11 +59,27 @@
...
 
 
-class DSSSourceRef(SourceRef[SimpleSourceName, 'DSSSourceRef']):
+@attr.s(frozen=True, auto_attribs=True, kw_only=True)
+class DSSSourceName(SourceName):
+    endpoint: str
+    prefix: str = config.dss_query_prefix
+
+    @classmethod
+    def parse(cls, name: str) -> SOURCE_NAME:
```
[the remark this answers is not in the document]
Comment (amarjandu) [reply src/azul/plugins/repository/dss/__init__.py:75]: The abstract method has to be implemented within this class. Unlike the TDRSourceName there is really nothing to extract from the endpoint that defines the source. I supposed that this method can accept a endpoint and return a instance of the class.

File: environment.py [hunk 16]
```diff
@@ -328,7 +328,7 @@ def env() -> Mapping[str, Optional[str]]:
         # and for subscriptions to the DSS. If this variable is set to a non-empty
         # string, only bundles whose UUID starts with the specified string will be
         # indexed.
-        'azul_dss_query_prefix': '',
+        'AZUL_DSS_QUERY_PREFIX': None,
```
[the remark this answers is not in the document]
Comment (amarjandu) [reply environment.py:331]: mistake on my part.

**Comment by amarjandu** [comment]
> However, since all sub-classes of `SourceName` have `prefix`, we _should_ actually pull it up into the base class. Since this is a fairly large change I don't think I need to see a fixup in the next round.

Ah I messed up the chained PR for this, modification for adding the `prefix` was included in `SourceName` was done in th chained PR. https://github.com/DataBiosphere/azul/pull/2931/commits/34d63f6f37ea5682aaca2c3b67cb3f74dbb6aaf2#r616017802 I'll bring in the `prefix` as a SourceName attribute within this PR.

I'm a little confoused about 
> I don't think I need to see a fixup in the next round.

I think that the appropriate step would be merging `Add prefix to SourceName` into `Add prefix to end of TDR Sources` then adding a `fixup` that addresses the comments from your last review. I'm going to hold of on doing this until I confirm with you.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
It appears that 

> However, since all sub-classes of SourceName have prefix, we should actually pull it up into the base class. Since this is a fairly large change I don't think I need to see a fixup in the next round.

from my previous review was not addressed. Still no fixup needed.

File: src/azul/indexer/document.py [hunk 17]
```diff
@@ -604,7 +604,7 @@ class DocumentSource(SourceRef[SimpleSourceName, SourceRef]):
 
     @classmethod
     def from_json(cls, source: JSON) -> 'DocumentSource':
-        return cls(id=source['id'], name=SimpleSourceName(source['name']))
+        return cls(id=source['id'], name=SimpleSourceName(name=source['name']))
```
Comment (amarjandu) [inline src/azul/indexer/document.py:607]: some funk here w/ using `name` so many times
Should be refactored to  `spec=SimpleSourceName(name=source['name'])` 
with changes from https://github.com/DataBiosphere/azul/issues/2843

File: src/azul/terra.py [hunk 18]
```diff
@@ -122,7 +130,8 @@ def parse(cls, source: str) -> 'TDRSourceName':
             is_snapshot = False
         else:
             assert False, type
-        self = cls(project=project, name=name, is_snapshot=is_snapshot)
+        validate_uuid_prefix(prefix)
+        self = cls(project=project, name=name, is_snapshot=is_snapshot, prefix=prefix)
```
Comment (hannes-ucsc) [inline src/azul/terra.py:134]: kwargs defined by the super class should come first.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: deployments/sandbox/environment.py [hunk 19]
```diff
@@ -53,19 +53,19 @@ def env() -> Mapping[str, Optional[str]]:
...
-            'tdr:broad-jade-dev-data:snapshot/hca_dev_20201203___20210219',
+            'tdr:broad-jade-dev-data:snapshot/hca_dev_20201203___20210219:42',
         ]),
         **{
             f'AZUL_TDR_{catalog.upper()}_SOURCES': ','.join([
-                'tdr:broad-jade-dev-data:snapshot/hca_dev_20201023_ebiv4___20210302'
+                f'tdr:broad-jade-dev-data:snapshot/hca_dev_20201023_ebiv4___20210302:{prefix}'
             ])
-            for catalog in ('dcp2ebi', 'it2ebi')
+            for catalog, prefix in (('dcp2ebi', '42'), ('it2ebi', ''))
```
[the remark this answers is not in the document]
Comment (amarjandu) [reply deployments/sandbox/environment.py:62]: @hannes-ucsc The IT will generate its own prefix at runtime,
https://github.com/DataBiosphere/azul/blob/28e48bb45eae0abaa4d9c4e9d72244cd4ea6e5d2/test/integration_test.py#L533
this prefix will be used to list_bundles then synth notifications. Currently when the list_bundles operation is called the listed bundles are NOT appended to any prefix that is defined in the configuration.

By adding the `source.name.prefix` from the TDRSourceName to the directly to the TDR query that is performed the prefix value that is generated from the IT is appended to the TDRSourceName prefix (`42`). This causes problems with the IT for some sources.


```
2021-04-27 09:07:59,573 INFO    MainThread: There are 3 bundle(s) with prefix '0' in source TDRSourceRef(id='c5cfdcda-995d-4f6e-94b7-e0532f6d83ca', name=TDRSourceName(prefix='42', project='broad-jade-dev-data', name='hca_dev_20201023_ebiv4___20210302', is_snapshot=True)).
2021-04-27 09:07:59,574 INFO    MainThread: Selecting 64 bundles with project metadata, out of 3 candidates, using random seed 3021656021161523122.
2021-04-27 09:07:59,574 INFO    MainThread: Not enough bundles with prefix '0' in catalog 'it2ebi'. Trying a shorter prefix.
2021-04-27 09:07:59,574 INFO    MainThread: Preparing notifications for catalog 'it2ebi' and prefix ''.
2021-04-27 09:07:59,574 INFO    MainThread: Listing bundles with prefix '' in source TDRSourceRef(id='c5cfdcda-995d-4f6e-94b7-e0532f6d83ca', name=TDRSourceName(prefix='42', project='broad-jade-dev-data', name='hca_dev_20201023_ebiv4___20210302', is_snapshot=True)).
2021-04-27 09:07:59,574 DEBUG   MainThread: Query: "\n            SELECT links_id, version\n            FROM broad-jade-dev-data.hca_dev_20201023_ebiv4___20210302.links\n            WHERE STARTS_WITH(links_id, '42')\n        "
2021-04-27 09:08:01,674 INFO    MainThread: There are 15 bundle(s) with prefix '' in source TDRSourceRef(id='c5cfdcda-995d-4f6e-94b7-e0532f6d83ca', name=TDRSourceName(prefix='42', project='broad-jade-dev-data', name='hca_dev_20201023_ebiv4___20210302', is_snapshot=True)).
2021-04-27 09:08:01,674 INFO    MainThread: Selecting 64 bundles with project metadata, out of 15 candidates, using random seed 3021656021161523122.
2021-04-27 09:08:01,675 WARNING MainThread: Not enough bundles in catalog 'it2ebi'. The test may fail.

2021-04-27 09:08:01,676 INFO    MainThread: Preparing notifications for catalog 'it3lungmap' and prefix 'bd'.
2021-04-27 09:08:02,161 INFO    MainThread: Listing bundles with prefix 'bd' in source TDRSourceRef(id='215793b0-b991-4b73-9584-1103d4072f39', name=TDRSourceName(prefix='42', project='broad-jade-dev-data', name='lungmap_dev_20210412__20210414', is_snapshot=True)).
2021-04-27 09:08:02,162 DEBUG   MainThread: Query: "\n            SELECT links_id, version\n            FROM broad-jade-dev-data.lungmap_dev_20210412__20210414.links\n            WHERE STARTS_WITH(links_id, '42bd')\n        "
2021-04-27 09:08:04,074 INFO    MainThread: There are 0 bundle(s) with prefix 'bd' in source TDRSourceRef(id='215793b0-b991-4b73-9584-1103d4072f39', name=TDRSourceName(prefix='42', project='broad-jade-dev-data', name='lungmap_dev_20210412__20210414', is_snapshot=True)).
2021-04-27 09:08:04,075 INFO    MainThread: Selecting 64 bundles with project metadata, out of 0 candidates, using random seed 3021656021161523122.
2021-04-27 09:08:04,075 INFO    MainThread: Not enough bundles with prefix 'bd' in catalog 'it3lungmap'. Trying a shorter prefix.
2021-04-27 09:08:04,075 INFO    MainThread: Preparing notifications for catalog 'it3lungmap' and prefix 'b'.
2021-04-27 09:08:04,075 INFO    MainThread: Listing bundles with prefix 'b' in source TDRSourceRef(id='215793b0-b991-4b73-9584-1103d4072f39', name=TDRSourceName(prefix='42', project='broad-jade-dev-data', name='lungmap_dev_20210412__20210414', is_snapshot=True)).
2021-04-27 09:08:04,075 DEBUG   MainThread: Query: "\n            SELECT links_id, version\n            FROM broad-jade-dev-data.lungmap_dev_20210412__20210414.links\n            WHERE STARTS_WITH(links_id, '42b')\n        "
2021-04-27 09:08:06,490 INFO    MainThread: There are 0 bundle(s) with prefix 'b' in source TDRSourceRef(id='215793b0-b991-4b73-9584-1103d4072f39', name=TDRSourceName(prefix='42', project='broad-jade-dev-data', name='lungmap_dev_20210412__20210414', is_snapshot=True)).
2021-04-27 09:08:06,490 INFO    MainThread: Selecting 64 bundles with project metadata, out of 0 candidates, using random seed 3021656021161523122.
2021-04-27 09:08:06,490 INFO    MainThread: Not enough bundles with prefix 'b' in catalog 'it3lungmap'. Trying a shorter prefix.
2021-04-27 09:08:06,490 INFO    MainThread: Preparing notifications for catalog 'it3lungmap' and prefix ''.
2021-04-27 09:08:06,490 INFO    MainThread: Listing bundles with prefix '' in source TDRSourceRef(id='215793b0-b991-4b73-9584-1103d4072f39', name=TDRSourceName(prefix='42', project='broad-jade-dev-data', name='lungmap_dev_20210412__20210414', is_snapshot=True)).
2021-04-27 09:08:06,490 DEBUG   MainThread: Query: "\n            SELECT links_id, version\n            FROM broad-jade-dev-data.lungmap_dev_20210412__20210414.links\n            WHERE STARTS_WITH(links_id, '42')\n        "
2021-04-27 09:08:09,485 INFO    MainThread: There are 0 bundle(s) with prefix '' in source TDRSourceRef(id='215793b0-b991-4b73-9584-1103d4072f39', name=TDRSourceName(prefix='42', project='broad-jade-dev-data', name='lungmap_dev_20210412__20210414', is_snapshot=True)).
2021-04-27 09:08:09,485 INFO    MainThread: Selecting 64 bundles with project metadata, out of 0 candidates, using random seed 3021656021161523122.
2021-04-27 09:08:09,485 WARNING MainThread: Not enough bundles in catalog 'it3lungmap'. The test may fail.
```


We might be able to figure out a prefix that fits our IT bundle requirements, but this can be tuned at a later time.

File: scripts/recan_bundle_tdr.py [hunk 20]
```diff
@@ -387,7 +387,8 @@ def main(argv):
         metadata = json.load(f)
 
     dss_source = DSSSourceRef(id='',
-                              name=SimpleSourceName(''))
+                              name=SimpleSourceName(name=config.dss_endpoint,
```
[the remark this answers is not in the document]
Comment (amarjandu) [reply scripts/recan_bundle_tdr.py:390]: Running `python scripts/recan_bundle_tdr.py` on the `develop` branch produces changes to
```
1b6d8348-d6e9-406a-aa6a-7ee886e52bf9.result.tdr.json
1b6d8348-d6e9-406a-aa6a-7ee886e52bf9.tables.tdr.json
```
running `python scripts/recan_bundle_tdr.py` on my branch also produces changes to the same files above.
Changing the value from `'42'` to `''` does not effect the changes to the files.

There is no diff between the changes that are produced between any test run.
```
(.venv) CBSE:azul amar$ diff 2650-canned-bundles/1b6d8348-d6e9-406a-aa6a-7ee886e52bf9.tables.tdr.json develop-canned-bundles/1b6d8348-d6e9-406a-aa6a-7ee886e52bf9.tables.tdr.json
(.venv) CBSE:azul amar$ diff 2650-canned-bundles/1b6d8348-d6e9-406a-aa6a-7ee886e52bf9.result.tdr.json develop-canned-bundles/1b6d8348-d6e9-406a-aa6a-7ee886e52bf9.result.tdr.json
(.venv) CBSE:azul amar$
```
[cans.zip](https://github.com/DataBiosphere/azul/files/6386518/cans.zip)

We can leave it as an empty string.

File: deployments/sandbox/environment.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply deployments/sandbox/environment.py:62]: The IT shortens the prefix until it finds enough bundles. If a configured prefix prevents the IT from finding any bundles, even with an IT prefix of `''` then the configured prefix is wrong and actually includes no bundles. So if a prefix doesn't work for a source of an IT catalog, that prefix wouldn't work for that source on a non-IT catalog either since the source wouldn't contribute any bundles to said catalog.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Comment by amarjandu** [comment]
add fixme: 
"add tooling to aid in prefix choice"
create issue for this as well.

**Comment by amarjandu** [comment]
IT failed during the 10min wait period for queue stabilization. All notifications were handled by the indexer lambdas, restarted IT to double check.

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
> IT failed during the 10min wait period for queue stabilization. All notifications were handled by the indexer lambdas, restarted IT to double check.

I didn't see this until now, after my approval. Looks like you asked for approval first, then ran IT. That would be out of order.

If the IT passes please assign to @achave11, move to Approved and check the two checklist items, please.

Otherwise, back to square one (apply fix fixes, ask for re-review).

**Comment by amarjandu** [comment]
> > IT failed during the 10min wait period for queue stabilization. All notifications were handled by the indexer lambdas, restarted IT to double check.
> 
> I didn't see this until now, after my approval. Looks like you asked for approval first, then ran IT. That would be out of order.
> 
> If the IT passes please assign to @achave11, move to Approved and check the two checklist items, please.
> 
> Otherwise, back to square one (apply fix fixes, ask for re-review).

Sorry that was for my reference, I'll keep those types of notes off the PR thread to reduce clutter going forward, the IT does pass on my local deployment.

**Comment by hannes-ucsc** [comment]
Just prefix them with "Note to self" or similar.
